src/main/java/org/folio/circulation/resources/PickSlipsReportResource.java

```diff
@@ -0,0 +1,199 @@
+  private static final String SERVICE_POINT_ID_PARAM = "servicePointId";
+  private static final String EFFECTIVE_LOCATION_ID_KEY = "effectiveLocationId";
+  private static final String PRIMARY_SERVICE_POINT_KEY = "primaryServicePoint";
+
+  private final String rootPath;
+
+  public PickSlipsReportResource(String rootPath, HttpClient client) {
+    super(client);
+    this.rootPath = rootPath;
+  }
+
+  @Override
+  public void register(Router router) {
+    RouteRegistration routeRegistration = new RouteRegistration(rootPath, router);
+    routeRegistration.getMany(this::getMany);
+  }
+
+  private void getMany(RoutingContext routingContext) {
+    final UUID servicePointId = UUID.fromString(routingContext.request().getParam(SERVICE_POINT_ID_PARAM));
+
+    final WebContext context = new WebContext(routingContext);
+    final Clients clients = Clients.create(context, client);
+
+    fetchLocationsForServicePoint(servicePointId, clients.locationsStorage())
+      .thenComposeAsync(r -> r.after(locations -> fetchPagedItemsByLocations(locations, clients.itemsStorage())))
+      .thenComposeAsync(r -> r.after(items -> filterItemsByOpenUnfilledRequests(items, clients.requestsStorage())))
+      .thenComposeAsync(r -> r.after(items -> fetchHoldings(items, clients.holdingsStorage())))
+      .thenComposeAsync(r -> r.after(items -> fetchInstances(items, clients.instancesStorage())))
+      .thenApply(r -> r.next(this::mapResultToJson))
+      .thenApply(OkJsonResponseResult::from)
+      .thenAccept(r -> r.writeTo(routingContext.response()));
+  }
+
+  private CompletableFuture<Result<Set<Location>>> fetchLocationsForServicePoint(
+    UUID servicePointId, CollectionResourceClient locationsStorageClient) {
+
+    return new MultipleRecordFetcher<>(locationsStorageClient, LOCATIONS_KEY, Location::from)
+      .findByQuery(exactMatch(PRIMARY_SERVICE_POINT_KEY, servicePointId.toString()))
+      .thenApply(r -> r.next(this::recordsToSet));
+  }
+
+  private CompletableFuture<Result<Set<Item>>> fetchPagedItemsByLocations(
+    Set<Location> locations, CollectionResourceClient itemStorageClient) {
+
+    Set<String> locationIds = locations.stream()
+      .map(Location::getId)
+      .collect(Collectors.toSet());
+
+    final Result<CqlQuery> statusQuery = exactMatch(STATUS_NAME_KEY, ItemStatus.PAGED.getValue());
+
+    return new MultipleRecordFetcher<>(itemStorageClient, ITEMS_KEY, Item::from)
+      .findByIndexNameAndQuery(locationIds, EFFECTIVE_LOCATION_ID_KEY, statusQuery)
+      .thenApply(r -> r.next(this::recordsToSet))
+      .thenApply(r -> r.next(items -> populateLocationsInItems(items, locations)));
+  }
+
+  private Result<Set<Item>> populateLocationsInItems(Set<Item> items, Set<Location> locations) {
+    return Result.succeeded(
+      items.stream()
+        .map(item -> item.withLocation(
+          locations.stream()
+            .filter(location -> location.getId().equals(item.getLocationId()))
```

Comment:
I'd create a Map with 'Location-Id'-'Location' pairs, to reduce complexity.

Reply:
Good idea. Done.